x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java

```diff
@@ -147,11 +147,27 @@ public PlanFactory visitEvalCommand(EsqlBaseParser.EvalCommandContext ctx) {
     public PlanFactory visitGrokCommand(EsqlBaseParser.GrokCommandContext ctx) {
         return p -> {
             String pattern = visitString(ctx.string()).fold().toString();
-            Grok result = new Grok(source(ctx), p, expression(ctx.primaryExpression()), Grok.pattern(source(ctx), pattern));
+            Grok.Parser grokParser = Grok.pattern(source(ctx), pattern);
+            validateGrokPattern(grokParser, pattern);
+            Grok result = new Grok(source(ctx), p, expression(ctx.primaryExpression()), grokParser);
             return result;
         };
     }
 
+    private void validateGrokPattern(Grok.Parser grokParser, String pattern) {
+        Map<String, DataType> definedAttributes = new HashMap<>();
+        for (Attribute field : grokParser.extractedFields()) {
+            String name = field.name();
+            DataType type = field.dataType();
+            DataType prev = definedAttributes.put(name, type);
+            if (prev != null) {
+                throw new ParsingException(
+                    "Invalid GROK pattern [" + pattern + "]: the attribute [" + name + "] is defined multiple times with different types"
```

Comment:
[A quick skim over comment] With EVAL we allow overwriting a field, irrespective of its type. Same with GROK(), when overriding input (i.e. already existing) fields. 
Wouldn't it make sense to allow this overriding, generally (i.e. even if different types)? Or, otherwise, reject overriding even if of the same type?

Reply:
Unfortunately the problem is more subtle here: for the calculation we rely on `org.elasticsearch.grok.Grok`, that is an ingest component, so we have limited control on how we parse the patterns and how we collect results.

Today we supports patterns like `%{WORD:foo} %{WORD:foo}` (attribute defined twice, same type) because GROK just returns a multi-value. Changing this behavior (ie. rejecting this pattern entirely or returning only one value) would be a breaking change. 

We could try to support some kind of override in case of different types, but I'm not sure it would be coherent with the behavior described above.

Reply:
This PR could be a temporary fix (a `ValidationException` is definitely better than an `IndexOutOfBoundException`) while we discuss and maybe implement a working alternative.